### PR TITLE
[MPS] Introduce torch.mps.Event() APIs

### PR DESCRIPTION
**Pull request by razarmehr**
- Implement `MPSEventPool` to recycle events.
- Implement python bindings with `torch.mps.Event` class using the MPSEventPool backend. The current member functions of the Event class are `record()`, `wait()`, `synchronize()`, `query()`, and `elapsed_time()`.
- Add API to measure elapsed time between two event recordings.
- Added documentation for Event class to `mps.rst`.
- Added test case to `test_mps.py`.

### CODE REVIEW

**Comment by razarmehr** [comment]
@albanD and @malfet , the CIs are green. Can you please review this PR? Thanks.

File: torch/mps/event.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    Args:
+        enable_timing (bool, optional): indicates if the event should measure time
+            (default: ``False``)
+    """
+
+    def __init__(self, enable_timing=False):
+        self.__eventId = torch._C._mps_acquireEvent(enable_timing)
+
+    def __del__(self):
+        # checks if torch._C is already destroyed
```
Comment (albanD) [inline torch/mps/event.py:18]: Did you actually encounter that?

File: test/test_mps.py [hunk 2]
```diff
@@ -7346,6 +7346,19 @@ def test_mps_profiler_module(self):
...
+        startEvent.record()
+        net1 = torch.nn.ConvTranspose2d(128, 64, kernel_size=3, stride=2, padding=1, output_padding=1)\
+            .to(device='mps', dtype=torch.float)
+        x = torch.rand(1, 128, 6, 6, device='mps', dtype=torch.float, requires_grad=True)
+        x = net1(x)
+        endEvent = torch.mps.Event(enable_timing=True)
+        endEvent.record()
+        elapsedTime = startEvent.elapsed_time(endEvent)
+        self.assertTrue(elapsedTime > 0.0)
+        print(f"Elapsed time of events: {elapsedTime:.3f} ms")
```
Comment (albanD) [inline test/test_mps.py:7360]: remove the print

**Review by albanD — APPROVED** [review]
Sorry for the delay in reviewing this.
The change sounds ok (I didn't check the mm implementation in detail as I assume this has already been reviewed in details).

I would say that the id-based binding is not the nicest though for two reasons: you have to `__del__` which have some caveats, the lifetime of the different objects are not hard-linked opening to bugs in having one dying earlier than the other and the binding code is not very easy to read.
If you plan on upgrading this, I would suggest to using pybind to directly bind the class via pybind like https://github.com/pytorch/pytorch/blob/18dacf7e793ec3ca3fb5c3d1e85485068e004d6e/torch/csrc/utils/init.cpp#L14  this will allow you to remove the need for the ID system and the del (as the python object can hold a strong reference to the c++ one via a shared ptr for example).

File: torch/mps/event.py (hunk 1 above)
Comment (razarmehr) [reply torch/mps/event.py:18]: Yes, I saw that error when torch._C got destroyed before our MPSEventPool instance.

File: test/test_mps.py (hunk 2 above)
Comment (razarmehr) [reply test/test_mps.py:7360]: Will do.

**Review by kulinseth — APPROVED** [review]

**Pull request closed without merging**